### PR TITLE
LPS-183427, LPS-183428 | Test Automation | Master

### PR DESCRIPTION
**Pull request by davisaints**
- **JIRA Tickets:**

[LPS-183427](https://issues.liferay.com/browse/LPS-183427)
[LPS-183428](https://issues.liferay.com/browse/LPS-183428)

### CODE REVIEW

**Comment by davisaints** [comment]
ci:test:sf

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase [hunk 1]
```diff
@@ -22,5 +22,63 @@ definition {
 	}
 
 
+	@description = "LPS-135398 - Verify unique values can be enforced for text fields"
+	@priority = 5
+	test CanPreventDuplicateText {
+		property custom.properties = "feature.flag.LPS-135398=true";
+		property portal.acceptance = "true";
+
+		task ("Given a custom object with a text field") {
```
Comment (andreefarias) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase:96]: Use the `task ("Given a custom object that contains a text field with unique values only')`

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase [hunk 2]
```diff
@@ -22,5 +22,63 @@ definition {
...
+			ObjectAdmin.addObjectFieldViaAPI(
+				fieldBusinessType = "Text",
+				fieldLabelName = "Custom Text Field",
+				fieldName = "customTextField",
+				fieldType = "String",
+				isRequired = "false",
+				objectName = "CustomObject183427");
+		}
+
+		task ("And given the text field contains a unique value only") {
```
Comment (andreefarias) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase:111]: This task is not necessary, you can use it together above.

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase [hunk 3]
```diff
@@ -22,5 +22,63 @@ definition {
 	}
 
 
+	@description = "LPS-135398 - Verify unique values can be enforced for text fields"
+	@priority = 5
+	test CanPreventDuplicateText {
+		property custom.properties = "feature.flag.LPS-135398=true";
```
Comment (andreefarias) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase:93]: Could remove this property? Today the FF was removed.

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase [hunk 4]
```diff
@@ -21,6 +21,71 @@ definition {
 		}
 	}
 
+	@description = "LPS-135398 - Verify unique values are enforced when editing an existing entry"
+	@priority = 5
+	test CanPreventDuplicateIntegerWhenEditing {
+		property custom.properties = "feature.flag.LPS-135398=true";
```
Comment (andreefarias) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase:27]: Could remove this property? Today the FF was removed.

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase [hunk 5]
```diff
@@ -21,6 +21,71 @@ definition {
...
+		property portal.acceptance = "true";
+
+		task ("Given a custom object with a integer field") {
```
Comment (andreefarias) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase:30]: Use the `task ("Given a Custom Object that contains an integer field with unique values only')`

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase [hunk 6]
```diff
@@ -21,6 +21,71 @@ definition {
...
+			ObjectAdmin.addObjectFieldViaAPI(
+				fieldBusinessType = "Integer",
+				fieldLabelName = "Custom Integer Field",
+				fieldName = "customIntegerField",
+				fieldType = "Integer",
+				isRequired = "false",
+				objectName = "CustomObject183428");
+		}
+
+		task ("And given the integer field contains a unique value only") {
```
Comment (andreefarias) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase:45]: This task is not necessary, you can use it together above.

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase [hunk 7]
```diff
@@ -21,6 +21,71 @@ definition {
...
+
+		task ("When the user attempts to edit an entry to match the value of the other entry") {
+			ObjectAdmin.goToCustomObject(objectName = "CustomObject183428");
+
+			ObjectPortlet.viewEntryDetails(entry = 15);
+
+			Type(
+				key_labelName = "Custom Integer Field",
+				locator1 = "TextInput#SPECIFIC_LABEL_TEXT_INPUT",
+				value1 = 30);
```
Comment (andreefarias) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase:80]: Could use the macro:  `ObjectPortlet.updateEntry`? It's more clear to read.

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase [hunk 8]
```diff
@@ -21,6 +21,71 @@ definition {
...
+
+			Button.clickSave();
```
Comment (andreefarias) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase:82]: Could use the macro: ObjectPortlet.updateEntry? It's more clear to read.

**Review by andreefarias — CHANGES_REQUESTED** [review]
Good job @davisaints! I requested only some small changes. Fix it, please.

**Review by andreefarias — APPROVED** [review]
Nice @davisaints! Thanks for the updates.

LGTM! 🎉

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+		var testPortalInstance = PropsUtil.get("test.portal.instance");
+
+		if (${testPortalInstance} == "true") {
+			PortalInstances.tearDownCP();
+		}
+	}
+
+
+
+}
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase]: You don't need to make a separate commit for this, you can just include this when adding the first test.

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase [hunk 10]
```diff
@@ -22,5 +22,60 @@ definition {
...
+			ObjectAdmin.goToCustomObject(objectName = "CustomObject183427");
+
+			ObjectAdmin.addObjectStringFieldEntryViaUI(
+				entryValue = "Entry Test",
+				invalidEntry = "true");
+		}
+
+		task ("Then an error is displayed and the entry isn't added") {
+			Alert.viewErrorMessage(errorMessage = "The Custom Text Field is already in use. Please enter a unique Custom Text Field.");
+		}
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase:136]: After checking for the alert, go back to the custom object and assert the object entry wasn't added. The main point of the test is to ensure we can't add the entry; checking for the alert is good but it doesn't actually prove the entry wasn't added.

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase [hunk 11]
```diff
@@ -21,6 +21,64 @@ definition {
...
+
+			ObjectPortlet.updateEntry(
+				entry = 12345,
+				invalidEntry = "true",
+				updateEntry = 54321);
+		}
+
+		task ("Then an error is displayed and the entry isn't updated") {
+			Alert.viewErrorMessage(errorMessage = "The Custom Integer Field is already in use. Please enter a unique Custom Integer Field.");
+		}
```
Comment (timpak) [inline modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase:80]: Similar to the previous test, we should go back and assert the entry wasn't updated by asserting the original entry is still present.

**Review by timpak — CHANGES_REQUESTED** [review]
@davisaints can you take a look at my comments below?

**Comment by davisaints** [comment]
ci:test:sf

File: modules/apps/object/object-test/src/testFunctional/paths/ObjectPortlet.path [hunk 12]
```diff
@@ -60,6 +60,11 @@
 		<td>//div[contains(@class,'dnd-tbody')]//*[contains(text(),'${key_entry}')]</td>
 		<td></td>
 	</tr>
+	<tr>
+		<td>ENTRY_VALUE_PER_ORDER</td>
+		<td>//div[contains(@class,'dnd-tbody')]//div[contains(@class, 'dnd-tr')][${key_order}]//*[contains(text(),'${key_entry}')]</td>
+		<td></td>
+	</tr>
```
Comment (davisaints) [inline modules/apps/object/object-test/src/testFunctional/paths/ObjectPortlet.path:67]: I noticed that we don't have something like this...

Comment (davisaints) [inline modules/apps/object/object-test/src/testFunctional/paths/ObjectPortlet.path:67]: We don't have a path for entries filtered by order

File: modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase (hunk 10 above)
Comment (davisaints) [reply modules/apps/object/object-test/src/testFunctional/tests/ObjectFieldsUniqueDate.testcase]: Ok, thank you!

**Review by timpak — CHANGES_REQUESTED** [review]
Hey @davisaints, can you take a look at my comments below?

File: modules/apps/object/object-test/src/testFunctional/paths/ObjectPortlet.path (hunk 12 above)
Comment (davisaints) [reply modules/apps/object/object-test/src/testFunctional/paths/ObjectPortlet.path:67]: I tried this way but it doesn't work, I need to filter by tbody and tr.

**Review by timpak — CHANGES_REQUESTED** [review]
@davisaints sorry, I missed this last time, can you change the file name to `ObjectFieldsUniqueValue`? I'm not sure why we're naming it as `ObjectFieldsUniqueDate`.

You can forward this PR once you make this change. Please do this in the original commits, don't do this in a separate commit.
